### ISSUE TITLE
[Question] hx-delete hits a 302 but will re-try with DELETE header causing an issue. Can I avoid retry 302 with the same method?

### ISSUE THREAD

**Issue by simkimsia** [bigskysoftware/htmx#322]
Initially I did this

```
<a class="btn btn-danger"
                    hx-confirm="Are you sure you want to delete this server?"
                    href="#" hx-delete="/servers/{{ server.id }}/" hx-target="body" role="button">Destroy</a></td>
```

And my backend will correctly delete and then redirect to `/servers`

However, the htmx will retry the 302 with DELETE. Thereby, triggering a 405. THis is because usually Django doesn't allow DELETE for list views.

SO I had to change the htmx to use post instead and also make sure the new url will correctly go to the same logic

```
<a class="btn btn-danger"
                    hx-confirm="Are you sure you want to delete this server?"
                    href="#" hx-post="/servers/{{ server.id }}/delete" hx-target="body" role="button">Destroy</a></td>
```

And this works because htmx will retry the 302 with POST. This time, Django backend will allow POST to list views

My question is can I avoid retry the 302 with the same method?

See my SO question and [comment](https://stackoverflow.com/questions/65654787/django-class-based-views-for-delete-method-cannot-redirect-to-listview?noredirect=1#comment116080922_65654787) where I thought the issue was initially Django related but wasn't

**Comment by 1cg** [comment]
That's strange, it should be following the standard Post/Redirect/Get logic as far as I know: 

https://en.wikipedia.org/wiki/Post/Redirect/Get

On the server side are you seeing a GET for the second request?  Or is it coming in as a delete?

Maybe there is a header that is being preserved across the requests for some reason?

**Comment by simkimsia** [comment]
coming in as delete for the second request. You can see the SO question where I reproduce the backend console.

go to https://stackoverflow.com/q/65654787/80353 and then scroll to **what it looks like in the console**

**Comment by 1cg** [comment]
I wonder if the difference is between the 302 and 303 redirect codes:

https://en.wikipedia.org/wiki/HTTP_302

https://en.wikipedia.org/wiki/HTTP_303

htmx just relies on the standard ajax functionality to handle redirects, so I don't think it's anything that htmx is doing explicitly.

**Comment by simkimsia** [comment]
Maybe add a test for 302 here to find out?

https://github.com/bigskysoftware/htmx/blob/fa92dc67ea065d06338a67a276d59bfdb9f85e7c/www/test/0.0.4/test/attributes/hx-delete.js

**Comment by 1cg** [comment]
Yeah, I should have a test there for sure.  However, all the tests use the sinon.js mock server:

https://sinonjs.org/releases/v9.2.3/fake-xhr-and-server/

And I doubt that the redirect logic is a good match for what the browser actually does.  I can write up a manual test with sinatra.rb...

**Comment by simkimsia** [comment]
anyway i can help?

**Comment by 1cg** [comment]
Could you use the network console in your browser to record the interaction and then post screen shots of the request headers and response headers?  And could you post the relevant server side code (just the part that issues the redirect)?

![image](https://user-images.githubusercontent.com/469183/104851496-87298400-58b2-11eb-9122-a2c8c0d88ca7.png)

**Comment by 1cg** [comment]
Hi there.

Did you have a chance to look at the requests in flight and see what they are returning?

**Comment by simkimsia** [comment]
> And could you post the relevant server side code (just the part that issues the redirect)?

I'm using the class-based views for Django 3.1 So the code is simply three lines.

```python
class ThingDetail(DeleteView):
    model = Thing
    success_url = reverse_lazy("things-list")
```

**Comment by simkimsia** [comment]
![CleanShot 2021-02-06 at 23 53 47](https://user-images.githubusercontent.com/245021/107123110-a9715900-68d6-11eb-9213-562c206c58c9.gif)

![image](https://user-images.githubusercontent.com/245021/107123119-b5f5b180-68d6-11eb-9a0d-19832394fa17.png)

![image](https://user-images.githubusercontent.com/245021/107123145-d6257080-68d6-11eb-8547-b2ebcadb3a99.png)

**Comment by simkimsia** [comment]
Right sorry ignore my last comment.

Should be this

![image](https://user-images.githubusercontent.com/245021/107136567-7b742f00-693f-11eb-9709-cc71f63a068d.png)

**Comment by simkimsia** [comment]
Ok i know how to solve this.

Tldr; I chose to use 303 at the server side so that it can correct redirect to the list view

Long story short is here https://stackoverflow.com/a/24375475/80353

In this SO answer which applies to Spring (a Java Framework), the question had the same issue as me.

1. Send a DELETE 
2. then server side want to redirect using 302
3. 302 will use precedent method and list typically don't accept DELETE as precedent method. Only POST, GET, and HEAD as precedent method

This seems like a web framework issue. But it's not. It appears to a convention most sensible web frameworks adopt.

There are 3 solutions with drawbacks:

## 1. override the convention

Allow the backend web framework to accept DELETE as precedent method for 302. 

Con: Not nice by convention

## 2. Let client handle redirection

send back a 200 then client will redirect back to list view

Con: This results in two requests and htmx-delete doesn't work that way. It will send a DELETE method request and then take whatever comes back and immediately swap. I like this so I want to keep this. One request to settle this rather than two.

## 3. Use 303 for the redirection

After successful delete, do a 303 redirect to list view (I chose this)

Con: 303 doesn't work with HTTP/1.0 and older browsers. But that's not a problem in the year 2021 and will continue to be less of a problem going forward.
